designs/use-selectors-at-cache.md

```diff
@@ -0,0 +1,117 @@
+specific fields it will still cache all the instances on that type increasing
+the memory and CPU load. This also get worse in case controller-runtime is
+deployed as a daemonset instead of a deployment since all the nodes with
+daemonset's pods will expect to receive notification of all the instances and
+that can end with scalation problems depending on the type or resource it's
+watching.
+
+As side effect of this proposal the request that does not match the filtered
+will not be returned, so users has to be warned that if this filtering
+is used they know it and they are careful with the request they are doing.
+
+The alternative that can be done with the current controller-runtme is
+to implement a custom cache that do the filtering since cache is pluggable
+into the controller-runtime [1]
+
+This proposal is related to the following issue [2]
+
+## Proposal
+
+Add a new selector code at `pkg/cache/selector/selector.go` with common structs
+and helpers
+```golang
+package selector
+
+...
+
+// SelectorsByGroupResource associate a GroupResource to a field/label selector
+type SelectorsByGroupResource map[schema.GroupResource]Selector
+
+// Selector specify the label/field selector to fill in ListOptions
+type Selector struct {
+	Label labels.Selector
+	Field fields.Selector
+}
+
+// FillInListOpts fill in ListOptions LabelSelector and FieldSelector if needed
+func (s Selector) FillInListOpts(listOpts *metav1.ListOptions) {
+...
+}
+
+// FindByGR use the GVR group and resource to find the field/label selector
+func (s SelectorsByGroupResource) FindByGR(gvr schema.GroupVersionResource) Selector {
+...
+}
+
+```
+
+Incrase `cache.Options` as follow:
+
+```golang
+type Options struct {
+	Scheme                  *runtime.Scheme
+	Mapper                  meta.RESTMapper
+	Resync                  *time.Duration
+	Namespace               string
+	SelectorByResource selector.SelectorsByGroupResource
+}
+```
+
+Add new builder function that will return a cache constructor that will return
+a cache filtered by selectors
+
+```golang
+// BuilderWithSelectors returns a Cache constructor that will filter
+// contents using selectors
+// WARNING: filtered out resources are not returned.
+func BuilderWithSelectors(selectors selector.SelectorsByGroupResource) NewCacheFunc {
+...
+}
+```
+
+is passed to informer's ListWatch and add the filtering option:
+
+```golang
+
+# At pkg/cache/internal/informers_map.go
+
+ListFunc: func(opts metav1.ListOptions) (runtime.Object, error) {
+            ip.selectors.FindByGR(mapping.Resource).FillInListOpts(&opts)
+...
+```
+
+Here is a PR with the implementatin at the `pkg/cache` part [3]
+
+## Example
+
+User will override `NewCache` function to make clear that they know exactly the
+implications of using a different cache than the default one
+
+```golang
+ ctrl.Options.NewCache = cache.BuilderWithSelectors(
+                            selector.SelectorsByGroupResource{
+							    {Group: "", Resource: "nodes"}: {
```

Comment:
I would probably go for `client.Object` instead of `schema.GroupResource` for configuring different caches (similar to how it is done in [`NewDelegatingClient`](https://github.com/kubernetes-sigs/controller-runtime/blob/eb0f1d4aa81794a38206965be25d118af5e18afc/pkg/client/split.go#L44).
It's more safe and consistent.
```suggestion
							    &corev1.Node: {
```

Reply:
Make sense, I there is a `apiutil.GVKForObject(obj, in.Client.Scheme())` to compose the GVK from a client.Object.

Reply:
The map for lookup can be built up once before hand, just like in the `delegatingReader`:
https://github.com/kubernetes-sigs/controller-runtime/blob/eb0f1d4aa81794a38206965be25d118af5e18afc/pkg/client/split.go#L46-L52

Anyways, this can be discussed in detail during the implementation review.

Reply:
I agree on the `client.Object` remark

Reply:
Done